### ISSUE TITLE
🐛 Biome encountered an unexpected error - AWS Amplify

### ISSUE THREAD

**Issue by ThaisRobba** [biomejs/biome#5820]
### Environment information

```block
CLI:
  Version:                      2.0.0-beta.2
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.14.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         yarn/1.22.22

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         /Users/thais/Documents/biome-sample/biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Workspace:
  Open Documents:               0
```

### What happened?

Running `yarn biome check biome.json` when there is aws-amplify (I know, it is super weird) as a project dependency results in the following:

```
$ /Users/thais/Documents/biome-sample/node_modules/.bin/biome check biome.json
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue.

When opening the issue, please provide a minimal reproduction, or identify and share the file/code that triggers it. Without a way to reproduce the error, the error can't be fixed:

Source Location: /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/node-semver-2.2.0/src/range.rs:982:18
Thread Name: biome::workspace_worker_7
Message: internal error: entered unreachable code: This should not have parsed: ~0.x.0

Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue.

When opening the issue, please provide a minimal reproduction, or identify and share the file/code that triggers it. Without a way to reproduce the error, the error can't be fixed:

Source Location: /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/node-semver-2.2.0/src/range.rs:982:18
Thread Name: main
Message: internal error: entered unreachable code: This should not have parsed: ~0.x.0

error Command failed with exit code 101.
```

### Expected result

Should lint as normal (beta 2.0.0-1 works fine).

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

### Reproduction repo

https://github.com/ThaisRobba/biome-repro-5820

**Comment by ematipico** [comment]
How does your package.json look like? One of our dependencies couldn't parse `~0.x.0`, which makes sense.

Please provide a reproduction so we can debug it.

**Comment by ThaisRobba** [comment]
Package.json for repro looks like this:

```
{
  "name": "origin",
  "version": "0.0.0",
  "license": "MIT",
  "scripts": {
  },
  "private": true,
  "dependencies": {
    "aws-amplify": "5.3.27"
  },
  "devDependencies": {
    "@biomejs/biome": "2.0.0-beta.2"
  }
}
```

**Comment by ThaisRobba** [comment]
Reproduction repo: https://github.com/ThaisRobba/biome-repro-5820

**Comment by ematipico** [comment]
It seems you didn't exclude `node_modules`. Please run `biome migrate` and try again

**Comment by TheKnarf** [comment]
I'm having the same problem on my project and we're not using `aws-amplify`.

Here is an repro `package.json`:

```
{
  "name": "biome-2-repro-5820",
  "version": "0.1.0",
  "private": true,
  "type": "module",
  "dependencies": {
  },
  "devDependencies": {
    "@biomejs/biome": "^2.0.0-beta.2",
    "@storybook/addon-actions": "^8.6.12",
    "@storybook/addon-essentials": "^8.6.12",
    "@storybook/addon-interactions": "^8.6.12",
    "@storybook/addon-links": "^8.6.12",
    "@storybook/addon-themes": "^8.6.12",
    "@storybook/nextjs": "^8.6.12",
    "@storybook/react": "^8.6.12",
    "@storybook/test": "^8.6.12",
    "@storybook/theming": "^8.6.12",
    "storybook": "^8.6.12",
    "storybook-dark-mode": "^4.0.2"
  },
  "packageManager": "pnpm@10.7.0+sha512.6b865ad4b62a1d9842b61d674a393903b871d9244954f652b8842c2b553c72176b278f64c463e52d40fff8aba385c235c8c9ecf5cc7de4fd78b8bb6d49633ab6"
}
```

**Comment by public** [comment]
Is there a workaround which fixes this on Beta 2? It's totally unusable currently.

**Comment by ematipico** [comment]
There isn't a workaround
